lecture_01/slice/slice.go

```diff
@@ -0,0 +1,253 @@
+	fmt.Printf("Slice5 after  rotating right by 1: %v\n", slice5)
+
+	// rotating right by 3
+	slice5 = rotateRight(slice5, 2)
+	fmt.Printf("Slice5 after  rotating right by 2: %v\n\n", slice5)
+
+	// get copy of slice5
+	slice6 := createCopy(slice5)
+	fmt.Printf("Slice6 after copying from slice5: %v\n", slice6)
+
+	// swapping odd and even elements of the slice
+	slice6 = swapElements(slice6)
+	slice7 := slice6
+	fmt.Printf("Slice6 after swapping odd and even elements: %v\n\n", slice6)
+
+	// sorting the slice in ascending order
+	fmt.Printf("Slice6 before: %v and ", slice6)
+	slice6 = helper.SortSlice(slice6, false)
+	fmt.Printf("after sorting in ascending  order: %v\n", slice6)
+
+	// sorting the slice in descending order
+	fmt.Printf("Slice7 before: %v and ", slice7)
+	slice7 = helper.SortSlice(slice7, true)
+	fmt.Printf("after sorting in descending order: %v\n\n", slice7)
+
+	// creating a new slice of strings
+	str := []string{"hello", "world", "core", "dev", "course"}
+	str2 := str
+
+	// sorting the string slice in ascending order using the same generic sort function
+	fmt.Printf("String slice before: %v and ", str)
+	str = helper.SortSlice(str, false)
+	fmt.Printf("after sorting in ascending  order: %v\n", str)
+
+	// sorting the string slice in descending order using the same generic sort function
+	fmt.Printf("String slice before: %v and ", str2)
+	str2 = helper.SortSlice(str2, true)
+	fmt.Printf("after sorting in descending order: %v\n", str2)
+}
+
+// addConstantToEachElement adds a constant value `x` to
+// each element of slice `s`. As we're just updating the
+// underlying array values, we don't need to return back s.
+func addConstantToEachElement(s []int, x int) {
+	for index := range s {
+		s[index] += x
+	}
+}
+
+// appendAtEnd appends a value `x` to the slice `s` at the end.
+func appendAtEnd(s []int, x int) []int {
+	s = append(s, x)
+	return s
+}
+
+// appendAtStart appends a value `x` to the slice `s` at the start.
+func appendAtStart(s []int, x int) []int {
+	if len(s) == 0 {
+		return append(s, x)
+	}
+
+	// append 2 slices which are s[0] and s[0:],
+	// which basically copies the 0th element twice
+	s = append(s[:1], s[0:]...)
+
+	// update the 0th element with `x`
+	s[0] = x
+
+	return s
+}
+
+// removeElement removes an element from the slice given an index.
+// It returns the removed element and slice.
+func removeElement(s []int, index int) (*int, []int) {
+	if len(s) == 0 && index >= 0 && index < len(s) {
+		return nil, nil
+	}
+
+	val := s[index]
+	s = append(s[:index], s[index+1:]...)
+
+	return &val, s
+}
+
+// appendAndRemoveDuplicates acceps 2 slices, appends them and
+// returns a new slice with unique elements.
+func appendAndRemoveDuplicates(s0 []int, s1 []int) []int {
+	// create an output slice to return
+	out := make([]int, 0, len(s0)+len(s1))
+
+	// create a map for keeping track of duplicates
+	m := make(map[int]bool)
+
+	for _, val := range s0 {
+		if _, ok := m[val]; !ok {
+			out = append(out, val)
```

Comment:
Or you could init out at the very end of the function with capacity of len(m), and fill it with elements of m.